mullvad-relay-selector/src/relay_selector/matcher.rs

```diff
@@ -0,0 +1,288 @@
+//! This module is responsible for filtering the whole relay list based on queries.
+use mullvad_types::{
+    constraints::{Constraint, Match},
+    custom_list::CustomListsSettings,
+    relay_constraints::{
+        BridgeState, InternalBridgeConstraints, Ownership, Providers, ResolvedLocationConstraint,
+        TransportPort,
+    },
+    relay_list::{
+        OpenVpnEndpoint, OpenVpnEndpointData, Relay, RelayEndpointData, WireguardEndpointData,
+    },
+};
+use talpid_types::net::{IpVersion, TransportProtocol, TunnelType};
+
+use super::query::{OpenVpnRelayQuery, RelayQuery, WireguardRelayQuery};
+
+/// The relay matcher decomposes a [`RelayQuery`] and exposes functions for filtering a set of relays
+/// to a subset which matches the original query.
+pub struct RelayMatcher<T: EndpointMatcher> {
+    /// Locations allowed to be picked from. In the case of custom lists this may be multiple
+    /// locations. In normal circumstances this contains only 1 location.
+    pub locations: Constraint<ResolvedLocationConstraint>,
+    /// Relay providers allowed to be picked from.
+    pub providers: Constraint<Providers>,
+    /// Relay ownership allowed to be picked from.
+    pub ownership: Constraint<Ownership>,
+    /// Concrete representation of [`RelayConstraints`] or [`BridgeConstraints`].
+    pub endpoint_matcher: T,
+}
+
+impl<'a> RelayMatcher<AnyTunnelMatcher<'a>> {
+    pub fn new(
+        query: RelayQuery,
+        openvpn_data: &'a OpenVpnEndpointData,
+        bridge_state: BridgeState,
+        wireguard_data: &'a WireguardEndpointData,
+        custom_lists: &CustomListsSettings,
+    ) -> RelayMatcher<AnyTunnelMatcher<'a>> {
+        let endpoint_matcher = AnyTunnelMatcher {
+            wireguard: WireguardMatcher::new(query.wireguard_constraints.clone(), wireguard_data),
+            openvpn: OpenVpnMatcher::new(
+                query.openvpn_constraints.clone(),
+                openvpn_data,
+                bridge_state,
+            ),
+            tunnel_type: query.tunnel_protocol,
+        };
+        Self::using(query, custom_lists, endpoint_matcher)
+    }
+}
+
+impl<T: EndpointMatcher> RelayMatcher<T> {
+    pub fn using(
+        query: RelayQuery,
+        custom_lists: &CustomListsSettings,
+        endpoint_matcher: T,
+    ) -> RelayMatcher<T> {
+        RelayMatcher {
+            locations: ResolvedLocationConstraint::from_constraint(query.location, custom_lists),
```

Comment:
This is not really a part of your PR, but here's a thought. The majority of references to `ResolvedLocationConstraint` are implementations of traits and convenience fns (done in `relay_selector.rs`), all actual uses are from the `matcher` module. It seems to me that we could reduce the complexity quite a bit by removing the `Location` variant of the enum, i.e. just make it a `Vec<GeographicLocationConstraint>` and just put a single location in the vector if not using custom lists.

Reply:
Done! It did indeed simplify things a bit. I suspect that there are more low-hanging fruit around, but this was a nice start!